Update CMSIS to 5.8.0

0xc0170 opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->

Update CMSIS to 5.8.0. It brings lot of fixes, see https://github.com/ARM-software/CMSIS_5/releases/tag/5.8.0

We again had some breakages, mainly it was arm compat (workaround implemented).

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [x] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

bulislaw committed:
CMSIS/RTX: Patch RTX4 to preserve osThreadDef compatibility

mbed OS used older RTX4 version and with osThreadDef accepting only 3
parameters, to preserve compatibility we hardcode the 'instances'
parameter to 1.

(cherry picked from commit 428acae1b2ac15c3ad523e8d40755a9301220822)
(cherry picked from commit 4360b7bbf815c4d812005938c9c27af199803a97)

bulislaw committed:
CMSIS/RTX: Allow overwriting mutex ops for ARMC

(cherry picked from commit 08ab8cc47d8722bf0c767990cd615cf1c427d006)

Commit:
CMSIS/RTX: Patch to conditionally compile

tz_context.c should be compiled only for secure world,
definition of API's in tz_context.h should be part of secure
binary/bootloader when building mbed-os as non-secure

(Cherry picked from d0a43b8af0eef4775ec4c3da5994ecceb9ed4558)

(cherry picked from commit fb354752eb69403ad503c8e53da67da6483776d6)

0xc0170 committed:
[CMSIS_5]: Updated to 13b9f72f2

0xc0170 committed:
cmsis: importer sha removal

I had to resolve conflicts with these sha. CMSIS 5.8.0 includes fixes and makes these changes
irrelevant.

0xc0170 commented on cmsis/CMSIS_5/CMSIS/RTOS2/RTX/Source/TOOLCHAIN_GCC/TARGET_M0/irq_cm0.S:
```diff
@@ -18,17 +18,21 @@
 
         .syntax  unified
 
+        #include "rtx_def.h"
```
this is causing a problem as we somehow get some C/C++ headers to this .S file and it just errors with strange errors. 

We ignored the inclusion previously because of this, will need to check how to fix it this time.

0xc0170 reviewed (commented):
I am able to compile a simple test now if I comment out rtx_def, so at least something

0xc0170 commented on cmsis/CMSIS_5/CMSIS/TARGET_CORTEX_M/Source/mbed_tz_context.c:
```diff
@@ -22,7 +22,7 @@
  */
 
-#if !FEATURE_TFM
+#if !TARGET_TFM
```
this one should be reverted as well, will check how it got in

0xc0170 replied:
removed

Patater commented on cmsis/CMSIS_5/CMSIS/RTOS2/RTX/Include/rtx_def.h:
```diff
@@ -0,0 +1,34 @@
+ * Title:       RTX derived definitions
+ *
+ * -----------------------------------------------------------------------------
+ */
```
This file failed CI due to missing SPDX header

0xc0170 replied:
I am reporting this upstream, it should be fixed asap. I assume as this is already released and new version might take some time, we can patch this file manually. I'll check.

0xc0170 replied:
Fix for upstream https://github.com/ARM-software/CMSIS_5/pull/1238

0xc0170 committed:
cmsis: fix license header in rtx_def

Upstream fix was merged https://github.com/ARM-software/CMSIS_5/pull/1238. Included here as it will be overwritten once again updated

0xc0170 committed:
retarget: move compat header for ARMClang prior any cmsis headers

Reference: CMSIS 5.8.0 known issues and https://github.com/ARM-software/CMSIS_5/issues/1211

This fixes the error about redefinition of enable/disable irq. we need compat header because of
semihosting (not yet provided in CMSIS).

0xc0170 committed:
semihosting: add compat header where is required

ARMCC provides __semihost via compat header. As CMSIS 5.8.0 removed this compat header,
we need to explicitly include it to fix definition missing error.

Commit:
cmsis: fix rtx_def inclusion in .S files for Gcc Arm

rtx_def includes two CMSIS headers that pull in C/C++ headers in our case. As I found out,
they should only define macros. We can fix it but it will require some refactoring as our targets
use mbed rtx headers to define heap using stdin header, plus some other offenders.

Workaround is to exclude the headers we do not need in irq assembly files.

Tracking issue https://github.com/ARMmbed/mbed-os/issues/14962

Commit:
cmsis: fix armcc compat header redefition of enable/disable irq

The fix will be in the 5.8.1 CMSIS, we cherry-pick it to our imported version.
See https://github.com/ARM-software/CMSIS_5/commit/e797cca3f42bef19a2c33b3b929f6010ddf53580

Commit:
musca targets: use IsException instead of removed isIrqMode

Commit:
cmsis: preprocess irq files

We use preprocessor for asm files even for Armcc. If symbol is defined it's replaced by preprocessor,
asembler would just see 1 or 0 in this case and errors:
TARGET_M33\\irq_armv8mml.S", line 31: Error: A1185E: Symbol missing

Use preprocessor instead.

Commit:
cypress: fix cy_syslib breakpoint undefined

CMSIS 5.8.0 requires for armcc to include compat header prior any other header.
See Known issues for the release.

Commit:
cmsis: patch fpu in asm for armcc5

For FPU, use armasm to select fpu selection. This will be fixed upstream in the next version of CMSIS. Meanwhile, we use our local patch.

Taken from --cpu selection for armasm:

--cpu           | {FPU}
Cortex-M4.fp.sp | VFPv4_SP_D16
Cortex-M7.fp.sp | FPv5-SP
Cortex-M7.fp.dp | FPv5_D16

Tracking issue: https://github.com/ARM-software/CMSIS_5/issues/1266

Commit:
rtos: fix delay with 0

Fixes error osErrorParameter that triggers our assert.
RTX delay ignored 0 previously, it now returns osErrorParameter instead.

Patater commented on platform/source/mbed_retarget.cpp:
```diff
@@ -53,6 +53,7 @@ struct DIR_impl {
 };
 
 #if defined(__ARMCC_VERSION)
+#   include <arm_compat.h>
```
A previous commit in this PR ("retarget: remove compat header file (ARMC5 old functions)
We should be using") said we should use the "newer safe implementation as explained in the release notes quoted below.". I'm a bit confused as to why we need to re-add this include back in. Could you explain in the commit message "Revert "retarget: remove compat header file (ARMC5 old functions)""?

Patater replied:
Maybe we only need the commit "retarget: move compat header for ARMClang prior any cmsis headers", and not these two commits that cancel each other out.

0xc0170 replied:
yes, I am rebasing now to remove them.

Patater commented on platform/source/LocalFileSystem.cpp:
```diff
@@ -14,6 +14,10 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
+#if defined(__ARMCC_VERSION)
```
In commit "semihosting: add compat header where is required",

```diff
- explictely
+ explicitly
```

0xc0170 replied:
fixed

Patater replied:
The spelling error in the commit message is still present.

Patater commented on platform/source/LocalFileSystem.cpp:
```diff
@@ -14,6 +14,10 @@
  * limitations under the License.
  */
+#if defined(__ARMCC_VERSION)
+#include <arm_compat.h>
```
In commit "semihosting: add compat header where is required", should this commit include the addition of this header to mbed_retarget.cpp as well? Shall we squash "semihosting: add compat header where is required" with "retarget: move compat header for ARMClang prior any cmsis headers"? The reason for both commits are the same, right?

Patater replied:
I don't see these commits were squashed yet.

Patater commented on cmsis/CMSIS_5/CMSIS/RTOS2/RTX/Source/TOOLCHAIN_GCC/TARGET_M0/irq_cm0.S:
```diff
@@ -26,6 +26,9 @@
 
         .syntax  unified
 
+        // Mbed OS patch: Exclude RTE_Components.h and RTX_Config.h inclusion in .S files
```
In commit message "cmsis: fix rtx_def inclusion in .S files for Gcc Arm", we say "they should only define macros.". Can we link to an open mbed-os issue to fix this?

0xc0170 replied:
fixed

Patater commented on cmsis/CMSIS_5/CMSIS/RTOS2/RTX/Source/TOOLCHAIN_ARM/TARGET_RTOS_M4_M7/irq_cm4f.S:
```diff
@@ -28,7 +28,7 @@
 #endif
 
-                IF       ({FPU}="FPv4-SP")
+                IF       ({FPU}="FPv4-SP") || ({FPU}="VFPv4_SP_D16") || ({FPU}="FPv5-SP") || ({FPU}="FPv5_D16")
```
"This will be fixed upstream in the next version of CMSIS." Could you provide a link as you did with other "fixed in next version" patches?

0xc0170 replied:
added tracking issue to the commit msg

Patater reviewed (dismissed):
LGTM. I guess remaining work is to update the patch list in cmsis_importer.json with final hashes.

Commit:
cmsis importer: add commits for Mbed OS changes

See specific SHA for details.
